### PR TITLE
RSDK-13904 Add module status to machine status

### PR DESCRIPTION
**Pull request by EvanDorsky**
This PR adds module status tracking to rdk and adds support for the new protos that module status tracking requires.

Scope (including the proto definition): [Module Status API](https://docs.google.com/document/d/1ahjQ-RW0ktGvTFFjtn66ANaYVV5vkeGyjS9qUWPXqdU/edit?tab=t.0#heading=h.tcicyojyqi6c)

Module status is exposed by `robot.MachineStatus` and is tracked by the module manager (`modmanager/manager.go`). `modmanager` has a new `Status` method that returns a list of module statuses for `robot.MachineStatus` to populate its response.

The module manager now tracks, for each module:
- `State` - the state of the module throughout its lifecycle
- `LastUpdated` - the time when a module was last updated
- `Error` - the error the caused the most recent transition into the `Unhealthy` state - this is cleared only when a module reenters the `Ready` state
- `ConsecutiveFailures` - the number of times a module has entered the `Unhealthy` state since the last time it entered the `Ready` state - this can detect some restart loops when a module goes back and forth between `Starting` and `Unhealthy`, but there's no cooldown logic for detecting (for example) modules that crash shortly after entering `Ready`

I also took this change as an opportunity to replace the existing failed modules tracking system, since a module being in the `Unhealthy` module state is equivalent to a module being in the `failedModules` list. I decided to rename a few methods since there weren't call sites in _that_ many files. Hopefully this doesn't cause headaches when merging but I'm also happy to revert the renames:

- `AddToFailedModules()` -> `SetModuleStatusUnhealthy(err) - this required a signature change in any case
- `FailedModules()` -> `UnhealthyModules()`
- `failedModules()` -> `unhealthyModules()`
- `ClearFailedModules()` removed, replaced with `PruneModuleStatuses(confs)` which addresses related but different needs

This can only merge after the protos in https://github.com/viamrobotics/api/pull/867 are merged and released

## Testing

`TestModuleStatus` in `module_lifecycle_test.go` tests all the possible module states across many different transitions and includes as many edge cases as I could think of. It also tests all the fields of the status struct - last updated time, error, consecutive failures.

I also added some module status checks in `TestCrashedModuleDependentRecovery`.

Tests pass locally but they won't pass in CI until the protos are merged.

### CODE REVIEW

**Review by allisonschiang — COMMENTED** [review]
only looked at the changes to failedModules but looks good to me!

File: module/modmanager/manager.go [hunk 1]
```diff
@@ -433,15 +450,10 @@ func (mgr *Manager) Reconfigure(ctx context.Context, conf config.Module) ([]reso
...
 
 	if err := mgr.startModule(ctx, mod); err != nil {
-		// could not start module during reconfiguration, add it to failedModules
-		mgr.AddToFailedModules(conf.Name)
 		// If re-addition fails, assume all handled resources are orphaned.
 		return handledResourceNames, err
 	}
 
-	// reconfiguration successful, remove from failed modules
-	mgr.deleteFromFailedModules(conf.Name)
```
Comment (aldenh-viam) [inline module/modmanager/manager.go:443]: Just wondering, why move these inside the `startModule` and `attemptRestart`? Seems like if someone adds a new exit point later but forgets a `mgr.SetModuleStatus*` before the return, the status could go out of sync.

File: module/status/status.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	ModuleStatePending
+	ModuleStateStarting
+	ModuleStateReady
+	ModuleStateUnhealthy
+	ModuleStateClosing
+)
+
+// Status represents the status of a module tracked by the module manager
+type Status struct {
+	Name                string
```
Comment (aldenh-viam) [inline module/status/status.go:23]: [nit] if `Status` is only ever stored in a map of Name -> Status and queried by name, you could skip some redundancy here.

**Review by aldenh-viam — APPROVED** [review]

File: robot/server/server.go [hunk 3]
```diff
@@ -557,6 +558,35 @@ func (s *Server) GetMachineStatus(ctx context.Context, _ *pb.GetMachineStatusReq
...
+	for _, modStatus := range mStatus.Modules {
+		pbModStatus := &pb.ModuleStatus{
+			ModuleName:          modStatus.Name,
+			LastUpdated:         timestamppb.New(modStatus.LastUpdated),
+			ConsecutiveFailures: uint32(modStatus.ConsecutiveFailures),
+		}
+		switch modStatus.State {
+		case modulestatus.ModuleStateUnknown:
+			// TODO: we might not want this to be an error
+			s.robot.Logger().CErrorw(ctx, "module in an unknown state", "module", modStatus.Name)
```
Comment (cheukt) [inline robot/server/server.go]: ya a warn is probably fine

File: module/modmanager/manager.go (hunk 1 above)
Comment (EvanDorsky) [reply module/modmanager/manager.go:443]: Good question - my reasoning for taking these calls out of Reconfigure was out of a desire to keep status update calls in as few functions as possible, and as close to the source of the transition as possible - hoping that this would minimize the likelihood that a new maintainer would have to add more status update calls.

Now that I'm looking at it, though, I ended up with many repetitive status update calls inside `startModule` (once per error, when the caller could just catch any error), and there are still status update calls in most of the callers of `startModule` and `startModuleProcess` (`Add`, `add`, and `attemptRestart`) anyway because there are other failures that they have to catch.

So maybe it would be better to move failure status updates up to the top level callers to minimize the amount of times those need to be written. I'll do that quickly and see if the tests still pass.

File: robot/impl/module_lifecycle_test.go [hunk 4]
```diff
@@ -960,3 +986,166 @@ func TestModuleAndResourceRemoval(t *testing.T) {
...
+	// Reconfigure in a goroutine so we can observe ModuleStartingState while the fake module is starting
+	reconfigureDone := make(chan struct{})
+	go func() {
+		defer close(reconfigureDone)
+		r.Reconfigure(ctx, cfg)
+	}()
+
+	testutils.WaitForAssertion(t, func(tb testing.TB) {
+		test.That(tb, p(getModuleStatus(tb, r, "fake").State), test.ShouldEqual, p(modulestatus.ModuleStateStarting))
+	})
```
Comment (cheukt) [inline robot/impl/module_lifecycle_test.go]: this will likely at some point flake (with tests that rely on timings, that's a chance that go scheduler will stop threads and restart them later, meaning that there will be cases where 500ms or even 5s will not be enough to observe the state)

could you wait for the assertion, then cancel the context to the Reconfigure

**Comment by cheukt** [comment]
generally looks good! but share the same concerns about having so many state transition calls

File: module/modmanager/manager.go (hunk 1 above)
Comment (cheukt) [reply module/modmanager/manager.go:443]: ya share the same concern here

Comment (EvanDorsky) [reply module/modmanager/manager.go:443]: I was able to get rid of like half a dozen status updates by raising the unhealthy status updates up to the higher level callers, I think this is definitely a maintainability/usability improvement, thanks for raising! Tests all pass, this will be up soon.

File: module/status/status.go (hunk 2 above)
Comment (EvanDorsky) [reply module/status/status.go:23]: It's actually stored in a list in the `MachineStatus` response to match how resource status is stored there, otherwise would totally agree.

File: robot/impl/module_lifecycle_test.go (hunk 4 above)
Comment (EvanDorsky) [reply robot/impl/module_lifecycle_test.go]: Hm I don't think I understand, are you suggesting to wait for the assertion indefinitely?

I wanted to observe the reconfigure -> starting -> unhealthy lifecycle, but I can see how this would flake (though I didn't see it over a bunch of local runs). I guess I observe the starting and unhealthy states in other places so could probably just remove this assertion.

File: module/modmanager/manager.go [hunk 5]
```diff
@@ -1156,34 +1164,106 @@ func getModuleDataParentDirectory(options modmanageroptions.Options) string {
 	return filepath.Join(options.ViamHomeDir, parentModuleDataFolderName, robotID)
 }
 
-// AddToFailedModules adds a failing module to the failedModules map.
-func (mgr *Manager) AddToFailedModules(moduleName string) {
-	mgr.failedModulesMu.Lock()
-	mgr.failedModules[moduleName] = true
-	mgr.failedModulesMu.Unlock()
+// These are the module status setters. Pending and Unhealthy are public
```
Comment (benjirewis) [inline module/modmanager/manager.go:1147]: [nit] You could probably put some of this "status" logic in a new file so it's more obviously distinct from the rest of manager code.

File: robot/impl/local_robot.go [hunk 6]
```diff
@@ -1849,10 +1854,6 @@ func (r *localRobot) reconfigure(ctx context.Context, newConfig *config.Config,
 		}
 	}()
 
-	if r.manager.moduleManager != nil {
```
Comment (benjirewis) [inline robot/impl/local_robot.go:1898]: Why were we doing this before? Clearing failed modules on every call to reconfigure?

File: robot/server/server.go [hunk 7]
```diff
@@ -557,6 +558,34 @@ func (s *Server) GetMachineStatus(ctx context.Context, _ *pb.GetMachineStatusReq
 
 		result.Resources = append(result.Resources, pbResStatus)
 	}
+	result.Modules = make([]*pb.ModuleStatus, 0, len(mStatus.Modules))
+	for _, modStatus := range mStatus.Modules {
```
Comment (benjirewis) [inline robot/server/server.go:579]: [opt] You could make (and unit-test) a function for this in `proto_conversions.go`.

File: robot/impl/local_robot.go (hunk 6 above)
Comment (EvanDorsky) [reply robot/impl/local_robot.go:1898]: I got the impression that this was because the original failed module tracking system couldn't remove failed modules if they left the failed state later - so I think the solution was to clear the list on reconfigure and let any failing modules re-populate themselves. But I'm not totally sure, I only looked into the old system far enough to make sure that I was replicating the intent when I built this.

File: robot/impl/module_lifecycle_test.go (hunk 4 above)
Comment (EvanDorsky) [reply robot/impl/module_lifecycle_test.go]: Spoke offline - I resolved this by adding a "slow start" capability to `testmodule` to guarantee a window of time when the module will be starting, regardless of how much time the reconfigure thread might be paused for. I also reduced the number of `WaitForAssertion` calls in the suite (a few were unnecessary, not sure why I put them there) and increased the time a great deal on the ones remaining (there are only two, and they each wait up to 5 minutes) so they're robust to long pauses.

File: module/status/status.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
+// Package modulestatus contains types describing the runtime status of a module
+// as tracked by the module manager. It is a leaf package so that both the module
+// manager and the robot package can depend on it without an import cycle.
+package modulestatus
```
Comment (cheukt) [inline module/status/status.go]: nit: can package just be part of "module"? moduleStatus.Status/State seems a bit clunky

File: robot/impl/module_lifecycle_test.go (hunk 4 above)
Comment (cheukt) [reply robot/impl/module_lifecycle_test.go]: looking at the changes here, I still think this will flake.

consider the case where the main test thread is paused at L1057 but the Reconfigure routine continues running. then Reconfigure can finish before we even start WaitForAssertion, and then the test will be stuck waiting for 5mins.

This does lead to cases for this test where there's not a whole we can do that is foolproof without rewriting a lot more code, but this change is worse than before (because there is a risk of us hanging for 5mins).

If we want to observe the starting state, I think it would be better to spin off the test separately and close the local robot after we observe the starting state. (I'm hoping that closing the local robot actually also shuts down the module manager's restart ctx, and subsequently kill the module process). this way we assert that the module does enter the starting state if the configuration includes it but nothing else. if that doesn't work I think we should have the slow startup and WaitForAssertionWithSleep both be around 15-20s

**Review by cheukt — CHANGES_REQUESTED** [review]
i think there is still a problem with the test, but moving the error setting up a level does look a lot cleaner! also you can rerequest reviews when the pr is ready again - i'm assuming it is because of standup/our convo earlier today

File: robot/impl/module_lifecycle_test.go (hunk 4 above)
Comment (EvanDorsky) [reply robot/impl/module_lifecycle_test.go]: Ah okay I see what you mean, when I rewrote the test this way I was only thinking about the goroutine pausing, not the main thread.

I think your original suggestion is better than what I did here and should address the new issue I introduced here: just go back to using `fakemodule` - which enters the starting state and never leaves - and then wait on the starting state assertion with a fairly long timeout, and cancel the context on the assertion. With `fakemodule` the main thread can never miss the starting state, and the assertion can only flake if the scheduler pauses the goroutine for longer than the assertion timeout, which can be made long enough.

File: module/status/status.go (hunk 8 above)
Comment (EvanDorsky) [reply module/status/status.go]: Ah yeah, that's a much better place to put it, will do that.

Comment (EvanDorsky) [reply module/status/status.go]: Oh, I might have misinterpreted you though if you were suggesting getting rid of the package and just defining `Status` within `module`. I actually originally wanted to do this, but importing `module` from `robot` was causing an import loop, and `robot` needs to know about module status because it defines `MachineStatus` which contains `ModuleStatus`.

I didn't like the semantics of this but it felt like the cleanest way to let everybody know about status was just to put it in its own package.

Comment (cheukt) [reply module/status/status.go]: ya that's fine then

**Comment by EvanDorsky** [comment]
I replaced the potentially flaky start/close waits in the test with new `testmodule` flags `VIAM_TESTMODULE_BLOCK_START` and `VIAM_TESTMODULE_BLOCK_CLOSE` that cause `testmodule` to wait on a socket while it's starting and closing. The test waits for the socket to open, asserts the state, and then closes - so it should be resilient to any pauses on either thread. Neither thread should be able to "miss" the other one, since if the reconfigure routine is paused, the main thread will wait on the socket connection indefinitely; and if the main thread is paused, the socket read waits indefinitely. Verified by adding some sleeps on each thread.

**Review by cheukt — APPROVED** [review]
lgtm % benjis suggestions - an extra unit test around the protoconversion would be good

**Comment by ghost** [comment]
Hey @EvanDorsky — this PR has been approved and CI has been green for 3+ business days. Ready to merge?


_Auto-comment from [overwatch](https://github.com/viamrobotics/overwatch). Will not re-nudge for 7 days._
